smithy-build/src/main/java/software/amazon/smithy/build/transforms/IncludeShapesBySelector.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.build.transforms;
+
+import java.util.Set;
+import software.amazon.smithy.build.TransformContext;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.selector.Selector;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.transform.ModelTransformer;
+
+/**
+ * {@code includeShapesBySelector} includes the shapes matching the given selector.
+ *
+ * <p>Prelude shapes are not removed by this transformer.
+ */
+public final class IncludeShapesBySelector extends BackwardCompatHelper<IncludeShapesBySelector.Config> {
+
+    /**
+     * {@code includeShapesBySelector} configuration.
+     */
+    public static final class Config {
+        private String selector = null;
+
+        /**
+         * Gets the selector used to filter the shapes.
+         *
+         * @return The selector used to filter the shapes.
+         */
+        public String getSelector() {
+            return selector;
+        }
+
+        /**
+         * Sets the selector used to filter the shapes.
+         *
+         * @param selector The selector used to filter the shapes.
+         */
+        public void setSelector(String selector) {
+            this.selector = selector;
+        }
+    }
+
+    @Override
+    public Class<Config> getConfigType() {
+        return Config.class;
+    }
+
+    @Override
+    public String getName() {
+        return "includeShapesBySelector";
+    }
+
+    @Override
+    String getBackwardCompatibleNameMapping() {
+        return "selector";
+    }
+
+    @Override
+    protected Model transformWithConfig(TransformContext context, Config config) {
+        String selector = config.getSelector();
+        ModelTransformer transformer = context.getTransformer();
+        Model model = context.getModel();
+        Set<Shape> selected = Selector.parse(selector).select(model);
+        return transformer.filterShapes(model, selected::contains);
```

Comment:
What happens if a selector returns, say, a list member? The list would be in an invalid state without the member. Does the ModelTransformer notice this, then remove the list, then remove any reference to the list?

Reply:
I don't think so but I can test, I wondered about a similar idea for the change to allow to filter enum members using tags and an explicit test to validate that an exception is thrown when all the members are removed was part of [that change](https://github.com/awslabs/smithy/blob/c45d5dbcc404c9ff3af6a08cc24c83377fc918b2/smithy-model/src/test/java/software/amazon/smithy/model/transform/RemoveShapesTest.java#L120).